File: ditto/api_models/router_ledger.py
"""Router-track score ledger published by the centralized router scorer.

The router track (SN118 "DittoBench" third competition) does **not** run its
eval on validators. One trusted ``dittobench-api`` scorer instance drives the
four third-party coding harnesses (Claude Code, Codex, opencode, Grok) against
each miner's router, computes a **correctness-floor-gated, token-cost-dominant**
score per miner, and publishes it as this ledger. Every validator only reads and
folds it — the same trust model as the memory ``LedgerResponse`` (each validator
folds one identical, scorer-decided number), so no provider secrets or heavy
harness containers ever touch a validator.

Field discipline mirrors ``ditto.api_models.validator.LedgerResponse``: unknown
fields are ignored (forward compatible), ``combined_score`` is the raw double the
scorer reported (never rounded, so every validator folds identical bytes), and an
absent/older feed simply yields no router entries — which folds to zero router
emission, exactly what the shadow ``weight_eligible=False`` state already does.
"""

from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class RouterHarness(StrEnum):
    """The big-four third-party coding harnesses the router must power.

    The string values are the stable ledger keys; the centralized scorer and the
    validator's failure classifier both key per-harness state on them.
    """

    CLAUDE_CODE = "claude_code"
    CODEX = "codex"
    OPENCODE = "opencode"
    GROK = "grok"


class RouterLedgerModel(BaseModel):
    model_config = ConfigDict(extra="ignore", frozen=True)


class RouterHarnessResult(RouterLedgerModel):
    """One harness's outcome for one miner router, as decided by the scorer.

    ``operational`` (the router powered the harness end-to-end) and ``floor_pass``
    (the run cleared the deterministic frontier-quality correctness floor) are the
    two gates; ``efficiency`` is the token-cost-dominant rank term in ``(0, 1]``,
    and ``upstream_token_cost_micros`` is the router's *upstream* provider token
    cost for the harness's tasks (informational — what compression bought). A
    harness contributes to the combined score only when both gates hold.
    """

    harness: RouterHarness
    operational: Annotated[
        bool,
        Field(description="The miner router powered this harness end-to-end."),
    ]
    floor_pass: Annotated[
        bool,
        Field(
            description=(
                "The run cleared the deterministic correctness floor at the "
                "frontier-quality bar (task build/tests). The gate, not the rank."
            )
        ),
    ]
    efficiency: Annotated[
        float,
        Field(
            ge=0.0,
            le=1.0,
            description=(
                "Token-cost-dominant efficiency score in [0, 1] the scorer "
                "assigned this harness slice (eff_h). Only meaningful when the "
                "harness was operational and cleared the floor."
            ),
        ),
    ]
    upstream_token_cost_micros: Annotated[
        int,
        Field(
            ge=0,
            description=(
                "Router's upstream provider token cost for this harness's tasks, "
                "in micro-units of the canonical cost model. Informational: the "
                "basis for the dominant token axis of eff_h."
            ),
        ),
    ]


class RouterLedgerEntry(RouterLedgerModel):
    """One miner's router-track result, published by the centralized scorer.

    ``combined_score`` is the scorer's soft, per-harness-weighted, floor-gated
    aggregate in ``[0, 1]`` (``Σ_h weight_h × (eff_h if operational and floor else
    0)``). A failed harness forfeits only its slice, so the weight destination is
    always this ``miner_hotkey`` and the fold ranks by this one number.
    """

    miner_hotkey: Annotated[str, Field(description="Miner's SS58 hotkey.")]
    agent_id: Annotated[UUID, Field(description="The miner's scored router agent.")]
    router_contract_version: Literal[1]
    weight_eligible: Annotated[
        bool,
        Field(
            description=(
                "Whether this router result may contribute emissions. False in "
                "shadow (v1). Tightened at promotion; the validator's track state "
                "is the authority, this is a defensive echo."
            )
        ),
    ]
    combined_score: Annotated[
        float,
        Field(
            ge=0.0,
            le=1.0,
            description=(
                "Soft per-harness-weighted, floor-gated aggregate in [0, 1]. The "
                "raw double the scorer reported (never rounded), the sole rank key."
            ),
        ),
    ]
    harnesses: Annotated[
        tuple[RouterHarnessResult, ...],
        Field(
            description=(
                "Per-harness outcomes (one entry per harness the scorer ran). "
                "Read by the validator's failure classifier for telemetry; the "
                "combined score already reflects any soft-forfeited slices."
            )
        ),
    ]
    first_seen: Annotated[
        datetime,
        Field(
            description=(
                "First-seen tie-break (UTC): when this miner's router lineage "
                "first reached the score it defends. The scorer resolves it; the "
                "validator folds it as served, so the original beats a later copy."
            )
        ),
    ]

    @field_validator("first_seen")
    @classmethod
    def first_seen_is_aware(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("router ledger first_seen must be timezone-aware")
        return value.astimezone(UTC)

    @model_validator(mode="after")
    def identifiers_are_nonzero(self) -> RouterLedgerEntry:
        if self.agent_id.int == 0:
            raise ValueError("router ledger agent_id is nil")
        if len({result.harness for result in self.harnesses}) != len(self.harnesses):
            raise ValueError("router ledger repeats a harness")
        return self


class RouterLedgerResponse(BaseModel):
    """Published by the centralized router scorer; read by every validator.

    Ordered highest-``combined_score`` first (ties broken by ``first_seen`` then
    ``agent_id``), the same deterministic order the router fold uses, so the
    exposed pool and the computed router weights agree by construction.
    """

    model_config = ConfigDict(extra="ignore")

    entries: list[RouterLedgerEntry] = Field(
        default_factory=list,
        description=(
            "One router result per miner, highest combined score first. "
            "Empty (or an older feed that omits it) folds to zero router "
            "emission, identical to the shadow state."
        ),
    )
    router_contract_version: Annotated[
        int | None,
        Field(
            default=None,
            ge=1,
            description=(
                "Router contract version this feed was scored under. None means "
                "the responding scorer predates the field."
            ),
        ),
    ] = None
    generated_at: Annotated[
        datetime | None,
        Field(default=None, description="When the router ledger was produced (UTC)."),
    ] = None
    stale: Annotated[
        bool,
        Field(
            default=False,
            description=(
                "True when the scorer served a last-known-good snapshot because a "
                "live read failed. The fold may still use it but should log it."
            ),
        ),
    ] = False
    count: Annotated[
        int, Field(default=0, ge=0, description="Number of entries returned.")
    ] = 0


__all__ = [
    "RouterHarness",
    "RouterHarnessResult",
    "RouterLedgerEntry",
    "RouterLedgerResponse",
]

File: services/dittobench-api/internal/routerledger/routerledger.go
// Package routerledger is the dittobench-api scorer's in-process accumulator for
// the SN118 router shadow ledger. Every submission's Dispatcher.Run produces one
// shadow LedgerEntry (internal/routerscore); this store keeps the latest entry
// per agent and serves a deterministically ordered snapshot to the publish route,
// which the platform relays to validators (who only read + fold it).
//
// Everything here is SHADOW: entries are never weight-eligible and their folded
// CombinedScore is 0. The store defensively refuses a weight-eligible entry so a
// future bug can never leak a weight-bearing number through the shadow publish
// path. Ordering is by the measured ShadowComposite (the dashboard number) so the
// exposed pool is meaningful even though it carries zero emission weight.
package routerledger

import (
	"sort"
	"sync"
	"time"

	"github.com/ditto-assistant/dittobench-api/internal/routerscore"
)

// Store accumulates the latest shadow ledger entry per agent, concurrency-safe.
type Store struct {
	mu          sync.RWMutex
	entries     map[string]routerscore.LedgerEntry
	generatedAt time.Time
}

// New builds an empty store.
func New() *Store {
	return &Store{entries: make(map[string]routerscore.LedgerEntry)}
}

// key selects the stable identity for an entry: agent_id when present, else the
// miner hotkey (shadow entries may carry only an agent id).
func key(entry routerscore.LedgerEntry) string {
	if entry.AgentID != "" {
		return entry.AgentID
	}
	return entry.MinerHotkey
}

// Record upserts a shadow entry. A weight-eligible entry is refused (shadow
// invariant); an entry with no identity is dropped. The earliest non-zero
// FirstSeen is preserved as the lineage tie-break so a later re-score cannot
// steal an original's first-seen precedence.
func (s *Store) Record(entry routerscore.LedgerEntry) {
	if entry.WeightEligible {
		return
	}
	id := key(entry)
	if id == "" {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if prev, ok := s.entries[id]; ok && !prev.FirstSeen.IsZero() {
		if entry.FirstSeen.IsZero() || prev.FirstSeen.Before(entry.FirstSeen) {
			entry.FirstSeen = prev.FirstSeen
		}
	}
	s.entries[id] = entry
	s.generatedAt = time.Now().UTC()
}

// Len reports how many agents are currently held.
func (s *Store) Len() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.entries)
}

// Snapshot returns a deterministically ordered ledger: highest ShadowComposite
// first, ties broken by earliest FirstSeen then agent_id. The returned slice is a
// copy, safe to marshal without holding the lock.
func (s *Store) Snapshot() routerscore.Ledger {
	s.mu.RLock()
	defer s.mu.RUnlock()
	entries := make([]routerscore.LedgerEntry, 0, len(s.entries))
	for _, e := range s.entries {
		entries = append(entries, e)
	}
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].ShadowComposite != entries[j].ShadowComposite {
			return entries[i].ShadowComposite > entries[j].ShadowComposite
		}
		if !entries[i].FirstSeen.Equal(entries[j].FirstSeen) {
			return entries[i].FirstSeen.Before(entries[j].FirstSeen)
		}
		return entries[i].AgentID < entries[j].AgentID
	})
	version := routerscore.RouterContractVersion
	ledger := routerscore.Ledger{
		Entries:               entries,
		RouterContractVersion: &version,
		Count:                 len(entries),
	}
	if !s.generatedAt.IsZero() {
		g := s.generatedAt
		ledger.GeneratedAt = &g
	}
	return ledger
}
